internal/impersonate/impersonate.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2020 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package impersonate is used to impersonate Google Credentials.
+package impersonate
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"golang.org/x/oauth2"
+)
+
+// Config for generating impersonated credentials.
+type Config struct {
+	// Target is the service account to impersonate. Required.
+	Target string
+	// Scopes the impersonated credential should have. Required.
+	Scopes []string
+	// Delegates are the service accounts in a delegation chain. Each service
+	// account must be granted roles/iam.serviceAccountTokenCreator on the next
+	// service account in the chain. Optional.
+	Delegates []string
+}
+
+// TokenSource returns an impersonated TokenSource configured with the provided
+// config using ts as the base credential provider for making requests.
+func TokenSource(ctx context.Context, ts oauth2.TokenSource, config *Config) (oauth2.TokenSource, error) {
+	if len(config.Scopes) == 0 {
+		return nil, fmt.Errorf("impersonate: scopes must be provided")
+	}
+	its := impersonatedTokenSource{
+		ctx:  ctx,
+		ts:   ts,
+		name: formatIAMServiceAccountName(config.Target),
+		// Default to the longest acceptable value of one hour as the token will
+		// be refreshed automatically.
+		lifetime: "3600s",
```

Comment:
coudl we make an optional "lifetime" config parameter:
 https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateAccessToken

the usecase is if i want to impersonate, then get an access token, then downscope/attenuate its permissions.   I can attenuate the resources the new token can access but can't toggle its lifetime.

if i can set the total lifetime on the parent token, that will get inherited to the downscoped token

see (using-impersonated-idtokens-or-impersonated-downscoped-credentials)[https://github.com/salrashid123/oauth2#using-impersonated-idtokens-or-impersonated-downscoped-credentials]

Reply:
This package is all internal right now is currently only exposed through the client option `WithImpersonatedCredentials`. There was some discussion in the design doc about having a configurable expiry, but we decided against it for now. We don't currently allow the lifetime of an of our tokens to be set. It was brought up that if we wanted to support expiry we would want to do it holistically with another client option. Also, this implementation auto-refreshes the token to make sure the TokenSource can be used w/o users having to create new impersonated credential.

If we did expose expiry, we would not want to refresh the token if set, right? Or do you think there is still valuable to have a short lifetime even though the token would be automatically refreshed? 

When/If we decide to make this package public I think such an option would be more useful.

Reply:
- if this is implemented all as an option, the use is directly with a GCP API then, correct?   Where i'm going at is there are common enough situations where i'd like derive other credential types source from impesonated credentials. For example, get an OIDC token that represents the impersonated credential.  Or use the impersonated credential to `signBlob()` (eg, generate [IAM based SignedURL](https://github.com/googleapis/google-cloud-go/issues/1130#issuecomment-668863950))

re: lifetime
* as mentioned, its exposed in the IAM today and has a current usecase (eg, derive a downscoped token)

* re: autorefresh.  this implmentation should chain back through with the underlying tokencsource/credentials once it expires if that is what is intended by the requestor. 

* all our other implementations supports this feature as a `Credential` by itsef that allows for setting the lifetime: [python](https://google-auth.readthedocs.io/en/latest/reference/google.auth.impersonated_credentials.html), [java](https://github.com/googleapis/google-auth-library-java/blob/master/oauth2_http/java/com/google/auth/oauth2/ImpersonatedCredentials.java#L107), [terraform](https://www.terraform.io/docs/providers/google/r/google_service_account.html)

i guess my biggest concern is the overall implementation (i.,e option vs as a tokensource in its own right).  Using the option fits in well w/ using this agains our APIs but there are other situations IAM based capabilities are used.

Reply:
> if this is implemented all as an option, the use is directly with a GCP API then, correct?

The primary use-case for exposing this as an option  is so it integrates nicely with any of our client libraries. For example:

```go
storage.NewClient(context.Background(), option. WithImpersonatedCredentials("foo@project.iam.gserviceaccount.com"))
```

But the option paradigm is used in other packages as well like [google.golang.org/api/transport](https://pkg.go.dev/google.golang.org/api/transport). This would allow someone to get a [*google.Credentials](https://pkg.go.dev/golang.org/x/oauth2/google?tab=doc#Credentials) from [Creds](https://pkg.go.dev/google.golang.org/api/transport?tab=doc#Creds); which would allow someone to get the impersonated [oauth2.TokenSource](https://pkg.go.dev/golang.org/x/oauth2@v0.0.0-20200107190931-bf48bf16ab8d?tab=doc#TokenSource)

> i guess my biggest concern is the overall implementation (i.,e option vs as a tokensource in its own right)

If we want to more easily empower other use-cases down the road we could look at refactoring the internal `impersonate` package and make it external. This would allow us the flexibility to also support other use cases like DwD more easily -- directly from that package.

Reply:
yeah, i can see how its convenient for the direct with a api library but its leaving out substantial usecases pointed out in the cited bug #378 as well as reasonable consistency with the other google auth library sets.

Can you provide a concrete working example of what was cited as deriving the `TokenSource` from this implementation?

Reply:
```go
        // import "google.golang.org/api/transport"
	creds, err := transport.Creds(context.Background(), option.WithImpersonatedCredentials("foo@project.iam.gserviceaccount.com"), option.WithScopes("https://www.googleapis.com/auth/cloud-platform"))
	if err != nil {
		log.Fatalf("problem getting creds: %v", err)
	}
	ts := creds.TokenSource
	_ = ts
```

Reply:
ok, thats better.. so the transport would allow you to chain options and auto-refresh them
 https://godoc.org/google.golang.org/api/option#WithCredentials
i.,e you provide a serviceaccountJSON and from there derive a credential/tokensource thats impersonated.  when the derived token expires, it will look towards the source.

i'm guessing ther'es another PR somewhere to attach a similar options to derived id_tokens
   https://godoc.org/google.golang.org/api/idtoken

Reply:
> ok, thats better.. so the transport would allow you to chain options and auto-refresh them
https://godoc.org/google.golang.org/api/option#WithCredentials

Correct.

> i'm guessing ther'es another PR somewhere to attach a similar options to derived id_tokens
https://godoc.org/google.golang.org/api/idtoken

Yes, some work will have to be done there to support the option. This bring up a good point though that we should return an error for now in the `idtoken` package until such a feature is implemented.

Reply:
ok, more thing,  if someone uses the `storage` library and its bootstrapped with the impersonated credentials objet, it will fail during signedURL operations because the library expects a real certificate to be present:

https://github.com/googleapis/google-cloud-go/blob/master/storage/storage.go#L640
  (i don't know if the storage client checks to see if the provided creds can even be used to sign though)

it gets a bit more convoluted but to support signedurl + impersonation requeres use of the IAM api to `signBlob()` again:

eg, like
  https://github.com/googleapis/google-cloud-go/issues/1130#issuecomment-668863950

for now i guess that can also throw an unsupported credential error or something

Reply:
> it will fail during signedURL operations because the library expects a real certificate to be present

The [SignedURL](https://pkg.go.dev/cloud.google.com/go/storage?tab=doc#SignedURL) stuff for storage is just a top level function that you pass options to. It does not use an authenticated client.

> to support signedurl + impersonation requeres use of the IAM api to signBlob()

Correct, in the case of impersonated credentials you would never have a PrivateKey anyways though, right? So I think you would need to take the approach of using `SignedURLOptions. SignBytes`.